# Data Cleaning with Pandas

In [1]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it with `brew install wget`, or use `curl <url> -o <filename>`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the file that you want to unzip.

In [2]:
sales_df = pd.read_csv('data/Real Property Sales.zip')

/usr/local/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
sales_df.head()

ExciseTaxNbr   Major Minor DocumentDate  SalePrice    RecordingNbr Volume  \
0       2687551  138860   110   08/21/2014     245000  20140828001436          
1       1235111  664885    40   07/09/1991          0  199203161090      071   
2       2704079  423943    50   10/11/2014          0  20141205000558          
3       2584094  403700   715   01/04/2013          0  20130110000910          
4       1056831  951120   900   04/20/1989      85000  198904260448      117   

  Page PlatNbr PlatType  ... PropertyType PrincipalUse SaleInstrument  \
0                        ...            3            6              3   
1  001  664885        C  ...            3            0             26   
2                        ...            3            6             15   
3                        ...            3            6             15   
4  053  951120        P  ...            3            0              2   

  AFForestLand  AFCurrentUseLand  AFNonProfitUse  AFHistoricProperty  \
0            N                 N               N                   N   
1            N                 N               N                   N   
2            N                 N               N                   N   
3            N                 N               N                   N   
4            N                 N               N                   N   

  SaleReason PropertyClass SaleWarning  
0          1             8              
1         18             3          11  
2         18             8    18 31 51  
3         11             8    18 31 38  
4          1             9          49  

[5 rows x 24 columns]

In [4]:
sales_df.describe()

ExciseTaxNbr     SalePrice  PropertyType  PrincipalUse  SaleInstrument  \
count  2.026164e+06  2.026164e+06  2.026164e+06  2.026164e+06    2.026164e+06   
mean   2.033068e+06  5.908315e+05  3.146618e+00  4.604457e+00    7.351473e+00   
std    5.633690e+05  5.994676e+06  3.831775e+00  2.619647e+00    6.613777e+00   
min    4.565830e+05 -6.000000e+02  0.000000e+00  0.000000e+00    0.000000e+00   
25%    1.576158e+06  0.000000e+00  3.000000e+00  2.000000e+00    3.000000e+00   
50%    2.044634e+06  1.500000e+05  3.000000e+00  6.000000e+00    3.000000e+00   
75%    2.511022e+06  3.495000e+05  3.000000e+00  6.000000e+00    1.500000e+01   
max    2.990860e+06  7.398850e+08  9.900000e+01  1.100000e+01    2.800000e+01   

         SaleReason  PropertyClass  
count  2.026164e+06   2.026164e+06  
mean   5.410099e+00   6.508585e+00  
std    6.238406e+00   2.484930e+00  
min    0.000000e+00   0.000000e+00  
25%    1.000000e+00   3.000000e+00  
50%    1.000000e+00   8.000000e+00  
75%    1.000000e+01   8.000000e+00  
max    1.900000e+01   9.000000e+00

### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [18]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [19]:
small_sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']].copy()

In [20]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.8+ MB


In [21]:
bldg_df = pd.read_csv('data/Residential Building.zip')

### Another warning! Which column has index 11?

In [22]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [23]:
bldg_df.head().T

0                           1  \
Major                                   62304                       62304   
Minor                                    9347                        9378   
BldgNbr                                     1                           1   
NbrLivingUnits                              1                           1   
Address              907  SW 104TH ST   98146  10215   11TH AVE SW  98146   
BuildingNumber                          907                         10215   
Fraction                                                                    
DirectionPrefix                            SW                               
StreetName          104TH                       11TH                        
StreetType                             ST                          AVE      
DirectionSuffix                                                        SW   
ZipCode                                 98146                       98146   
Stories                                     1                           1   
BldgGrade                                   7                           7   
BldgGradeVar                                0                           0   
SqFt1stFloor                             1180                         970   
SqFtHalfFloor                               0                           0   
SqFt2ndFloor                                0                           0   
SqFtUpperFloor                              0                           0   
SqFtUnfinFull                               0                           0   
SqFtUnfinHalf                               0                           0   
SqFtTotLiving                            1570                         970   
SqFtTotBasement                           670                           0   
SqFtFinBasement                           390                           0   
FinBasementGrade                            6                           0   
SqFtGarageBasement                          0                           0   
SqFtGarageAttached                          0                         380   
DaylightBasement                                                        N   
SqFtOpenPorch                             150                           0   
SqFtEnclosedPorch                           0                          80   
SqFtDeck                                    0                           0   
HeatSystem                                  5                           5   
HeatSource                                  2                           2   
BrickStone                                  0                           0   
ViewUtilization                                                         N   
Bedrooms                                    3                           3   
BathHalfCount                               0                           0   
Bath3qtrCount                               0                           1   
BathFullCount                               1                           1   
FpSingleStory                               1                           1   
FpMultiStory                                0                           0   
FpFreestanding                              0                           0   
FpAdditional                                0                           0   
YrBuilt                                  1965                        1969   
YrRenovated                                 0                           0   
PcntComplete                                0                           0   
Obsolescence                                0                           0   
PcntNetCondition                            0                           0   
Condition                                   3                           3   
AddnlCost                                   0                           0   

                                            2                           3  \
Major                                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [24]:
small_bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']].copy()

In [25]:
small_bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 512114 entries, 0 to 512113
Data columns (total 4 columns):
Major            512114 non-null int64
Minor            512114 non-null int64
SqFtTotLiving    512114 non-null int64
ZipCode          468423 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [26]:
sales_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

In [14]:
pd.merge?

In [27]:
sales_data.head()

Major Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860   110   08/21/2014     245000           1490   98002
1  138860   110   06/12/1989     109300           1490   98002
2  138860   110   01/16/2005      14684           1490   98002
3  138860   110   06/08/2005          0           1490   98002
4  423943    50   10/11/2014          0            960   98092

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [28]:
pd.to_numeric(sales_df['Major'])

ValueError: Unable to parse string "      " at position 949922

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [29]:
# The single question mark means "show me the docstring"
pd.to_numeric?

Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [30]:
small_sales_df.loc[:,'Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [31]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.8+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [32]:
small_sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [33]:
small_sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2026164 entries, 0 to 2026163
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.8+ MB


Now, let's try our join again.

In [34]:
sales_data = pd.merge(small_sales_df, small_bldg_df, on=['Major', 'Minor'])

In [35]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [36]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1446234 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1446234 non-null float64
Minor            1446234 non-null float64
DocumentDate     1446234 non-null object
SalePrice        1446234 non-null int64
SqFtTotLiving    1446234 non-null int64
ZipCode          1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 77.2+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [37]:
sales_data.loc[sales_data['ZipCode'].isna()].head(20)

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
91   334330.0  1343.0   05/30/2006          0           4600     NaN
92   334330.0  1343.0   05/30/2006          0           4600     NaN
93   334330.0  1343.0   11/26/2001          0           4600     NaN
94   334330.0  1343.0   05/30/2006          0           4600     NaN
95   334330.0  1343.0   06/30/2016          0           4600     NaN
96   334330.0  1343.0   05/30/2006          0           4600     NaN
97   334330.0  1343.0   08/18/2005          0           4600     NaN
98   334330.0  1343.0   06/07/2007     915000           4600     NaN
99   334330.0  1343.0   11/26/2001     403000           4600     NaN
100  334330.0  1343.0   11/20/2014     880000           4600     NaN
101  334330.0  1343.0   05/30/2006          0           4600     NaN
102  663990.0    40.0   11/16/2006     690576           2540     NaN
103  663990.0    40.0   08/31/2005     480000           2540     NaN
104  663990.0    40.0   02/24/2006          0           2540     NaN
105  663990.0    40.0   02/22/1989          0           2540     NaN
106  663990.0    40.0   06/24/2015     800000           2540     NaN
107   32103.0   230.0   12/16/2009    2340000           2300     NaN
108   32103.0   230.0   12/09/2009          0           2300     NaN
109   32103.0   230.0   06/18/2012     333370           2300     NaN
110   32103.0   240.0   12/16/2009    2340000           2750     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [38]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]

In [39]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1329081 entries, 0 to 1446233
Data columns (total 6 columns):
Major            1329081 non-null float64
Minor            1329081 non-null float64
DocumentDate     1329081 non-null object
SalePrice        1329081 non-null int64
SqFtTotLiving    1329081 non-null int64
ZipCode          1329081 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 71.0+ MB


# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [40]:
sales_data = sales_data.loc[(sales_data['SalePrice'] > 0) & sales_data['SqFtTotLiving'] > 0]
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
294  935840.0  210.0   06/10/2005    2025000           4311   98003
295  935840.0  210.0   11/17/2006     910000           4311   98003
296  935840.0  210.0   03/27/2006     258000           4311   98003
301  935840.0  300.0   06/10/2005    2025000           3497   98092
302  935840.0  300.0   03/08/2012     554300           3497   98092

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [41]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

pd.to_numeric(sales_data['ZipCode'])
# sales_data.ZipCode.value_counts()

294        98003.0
295        98003.0
296        98003.0
301        98092.0
302        98092.0
303        98092.0
340        98198.0
444        98053.0
445        98053.0
446        98053.0
984        98188.0
2909       98001.0
2910       98001.0
2911       98001.0
3631       98010.0
3632       98010.0
3633       98010.0
3797       98034.0
3798       98034.0
3799       98034.0
4244       98001.0
4245       98001.0
4545       98199.0
4548       98199.0
4549       98199.0
5298       98177.0
5299       98177.0
5301       98177.0
5964       98075.0
5965       98075.0
            ...   
1441778    98023.0
1441827    98045.0
1441828    98045.0
1441918    98033.0
1442067    98198.0
1442294    98001.0
1442384    98039.0
1442478    98004.0
1442677    98074.0
1442765    98022.0
1442971    98092.0
1442972    98092.0
1443254    98032.0
1443876    98118.0
1444044    98030.0
1444171    98031.0
1444279    98166.0
1444329    98001.0
1444423    98052.0
1444440    98004.0
1444614    98177.0
1444905    9

### 3. Add a column for PricePerSqFt



### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!